src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -177,16 +177,31 @@ bool TryInlineBodyDependencies (ref BodyReducer reducer)
 					if (!constExprMethods.TryGetValue (md, out targetResult))
 						break;
 
+					bool explicitlyAnnotated = Annotations.GetAction (md) == MethodAction.ConvertToStub;
+
 					// Allow inlining results of instance methods which are explicitly annotated
 					// but don't allow inling results of any other instance method.
 					// See https://github.com/mono/linker/issues/1243 for discussion as to why.
 					// Also explicitly prevent inlining results of virtual methods.
 					if (!md.IsStatic &&
-						(md.IsVirtual || Annotations.GetAction (md) != MethodAction.ConvertToStub))
+						(md.IsVirtual || !explicitlyAnnotated))
 						break;
 
-					if (md.HasParameters)
-						break;
+					// Allow inlining results of methods with by-value parameters which are explicitly annotated
+					// but don't allow inlining of results of any other method with parameters.
+					if (md.HasParameters) {
+						if (!explicitlyAnnotated)
+							break;
+
+						bool hasByRefOrOutParameter = false;
+						for (int paramIndex = 0; paramIndex < md.Parameters.Count; paramIndex++) {
+							var param = md.Parameters[paramIndex];
+							hasByRefOrOutParameter |= param.ParameterType.IsByReference || param.IsOut;
```

Comment:
If you specifically want `out` in the C# sense, this should be `param.ParameterType.IsByReference && param.IsOut`, but maybe we could just focus on the IsByReference part.

(This:

```csharp
static void Blah([Out] int x) { }
```

is just some harmless odd piece of metadata.)